### PR TITLE
contracts: Use `proof_size` from benchmarks

### PR DESCRIPTION
**Pull request by athei**
We remove our ad-hoc injected `proof_size` and replace it by the benchmarking results. We also refactor `wasm::code_cache::store`  to never load the code from storage if it was supplied via the extrinsic.

We use `#[pov_mode = Measure]` everywhere because the max encoded length of the used storage items is usually much bigger than the average. All benchmarks are already written in a way that the `proof_size` scales with the components just as the `ref_time` does.

We also fixed the deletion queue which wasn't honouring the `proof_size`. In the process we added a new associated function: `Weight::checked_div_per_component`.

### CODE REVIEW

**Comment by athei** [comment]
/cmd queue -c bench $ pallet runtime pallet_contracts

**Comment by athei** [comment]
/cmd queue -c bench $ pallet dev pallet_contracts

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2331436

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2334261

File: frame/contracts/src/weights.rs
Comment (athei) [inline frame/contracts/src/weights.rs:253]: Oof. `5` bytes per byte of code. That is bad. It is probably due to our instrumentation worst case. We really need to get rid of that and move that into wasmi.

File: frame/contracts/src/wasm/code_cache.rs
Comment (xermicus) [inline frame/contracts/src/wasm/code_cache.rs:180]: Why is this changed?

Comment (athei) [reply frame/contracts/src/wasm/code_cache.rs:180]: We were overestimating the size of the code by more than we need to. By that point we already know the size of the instrumented code and can use this as the upper bound. Then after re-instrument we know the size of the pristine code which the benchmarks are based upon.

Comment (xermicus) [reply frame/contracts/src/wasm/code_cache.rs:180]: I see, no way the orig code size will ever be bigger than the instrumented one

**Review by xermicus — COMMENTED** [review]
LGTM

**Review by xermicus — APPROVED** [review]

**Comment by athei** [comment]
bot rebase

File: frame/contracts/src/wasm/code_cache.rs
Comment (athei) [reply frame/contracts/src/wasm/code_cache.rs:180]: Exactly. It does not remove instructions. Only adds.

**Review by agryaznov — CHANGES_REQUESTED** [review]
Please consider the suggested small fixes to the schedule. 

Also, I don't get this sentence from the PR description:

> We also refactor wasm::code_cache::store to never load the code from storage if it was supplied via the extrinsic. 

Do you mean that: it was refactored "_not to write the contract code supplied via extrinsic to storage, as it is already done eagerly after re-instrumentation_"

File: frame/contracts/src/schedule.rs
Comment (athei) [reply frame/contracts/src/schedule.rs]: Good catch. The number is still small. It makes sense. We need to touch a contract's account anyways on instantiate so sending some value there doesn't make a big difference. On `call` we don't touch the account unless we are making a transfer.

File: frame/contracts/src/schedule.rs
Comment (athei) [reply frame/contracts/src/schedule.rs]: Good catch.

**Review by agryaznov — APPROVED** [review]

**Comment by athei** [comment]
Don't merge until we re-benchmark on the correct machine.

**Comment by agryaznov** [comment]
looks like we're waiting new weights from here: https://github.com/paritytech/substrate/pull/13316

**Comment by athei** [comment]
> looks like we're waiting for new weights from here: #13316

Yes. It makes no sense to benchmark before this is merged.

**Comment by athei** [comment]
/cmd queue -c bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev <pallet_name>

**Comment by athei** [comment]
/cmd queue -c bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Comment by athei** [comment]
/cmd queue -c bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Comment by mordamax** [comment]
bot bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Comment by alvicsam** [comment]
bot bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Comment by alvicsam** [comment]
bot bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Comment by alvicsam** [comment]
bot cancel 4-ab624254-7a25-4d78-b3eb-acac65c93cdb

**Comment by alvicsam** [comment]
bot bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Comment by alvicsam** [comment]
bot bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Comment by alvicsam** [comment]
I've updated rust to `1.66.1` and nightly to `nightly-2022-11-16` on the runner. But now it looks that something broken in the code: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2367880#L224

**Comment by athei** [comment]
How should something break in the code without the code being changed? The CI is green on this PR. Those are linker errors. Probably some libc or cache errors.

**Comment by alvicsam** [comment]
I'm not talking about this PR. The problem is currently in master. I tried running `cargo build --profile=production --locked --features=runtime-benchmarks` in current ci image and had the same problem (there was no cache)

**Comment by athei** [comment]
> I'm not talking about this PR. The problem is currently in master. I tried running `cargo build --profile=production --locked --features=runtime-benchmarks` in current ci image and had the same problem (there was no cache)

Is there an issue for that?

**Comment by alvicsam** [comment]
Yes, both old and new runners are affected. @ggwpez recommended me to use `git bisect` to find the commit and, yes, https://github.com/paritytech/substrate/pull/13160 is the first one where `cargo build --profile=production --locked --features=runtime-benchmarks` fails

**Comment by athei** [comment]
> Yes, both old and new runners are affected. @ggwpez recommended me to use `git bisect` to find the commit and, yes, #13160 is the first one where `cargo build --profile=production --locked --features=runtime-benchmarks` fails

Yeah no bisect necessary as we already found out which commit broke it: https://github.com/paritytech/substrate/issues/13376

**Comment by athei** [comment]
bot clean

**Comment by athei** [comment]
bot bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2392603

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2392925

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2392924

File: frame/contracts/src/weights.rs
Comment (athei) [inline frame/contracts/src/weights.rs:200]: This completely brokes down as if no work is performed anymore. Neither does `k` add a write (which is should) nor is there a substantial amount of weight attached to `k`. Need to investigate. This is not the case on master.

**Comment by athei** [comment]
bot bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Pull request merged**
